plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java

```diff
@@ -151,9 +157,23 @@ protected void runOCBinary() {
 				return portForwarding;
 			}
 		}, null);
-		if (portForwarding != null) {
-			REGISTRY.put(pod, portForwarding);
-		}
+        if (portForwarding != null) {
+            REGISTRY.put(pod, portForwarding);
+            IConnectionsRegistryListener podListener = new ConnectionsRegistryAdapter() {
+                @Override
+                public void connectionChanged(IConnection connection, String property, Object oldValue, Object newValue) {
+                    if (newValue == null && oldValue instanceof IPod) {
```

Comment:
Where do you check that the deleted pod is the one you started a port forward ?

Reply:
I guess something like ```REGISTRY.containsKey(pod)``` is missing.
@bdshadow WDYT?

Reply:
the following does the trick for me:
```
+++ b/plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java
@@ -162,7 +162,9 @@ public class PortForwardingUtils {
             IConnectionsRegistryListener podListener = new ConnectionsRegistryAdapter() {
                 @Override
                 public void connectionChanged(IConnection connection, String property, Object oldValue, Object newValue) {
-                    if (newValue == null && oldValue instanceof IPod) {
+                    if (newValue == null
+                               && oldValue instanceof IPod
+                               && REGISTRY.containsKey(oldValue)) {
                         try {
```

steps:
1) ASSERT: have 2 applications in 2 different projects with at least 1 pod runnin
2) EXEC: start port forwarding for both pods (2nd one needs to be started by checking "[x] find free local ports"
3) EXEC: in cmd line, verify that 2 "oc port-forwarding" processes (see exact semantic above) are running
4) EXEC: kill 1 pod
5) ASSERT: the port forwarding for the pod, that you just killed, was terminated, the other port port forwarding is still running

Reply:
@jeffmaury @adietish if you look inside `PortForwardingUtils.stopPortForwarding((IPod)oldValue, null);`, you'll find that first of all there is a check `PortForwardingUtils.isPortForwardingStarted(pod)`, which actually checks if a pod was put into `REGISTRY` before

Reply:
@adietish checked your steps. Everything works for me without additional check ` && REGISTRY.containsKey(oldValue)`. As i said above, this check already exists inside `PortForwardingUtils.isPortForwardingStarted(pod)`

Reply:
@bdshadow yes I think that you're right. Resolving. @jeffmaury dare to comment please?

Reply:
+1